src/vm/mod.rs

```diff
@@ -48,6 +48,7 @@ where
 
     /// The type of edges in this VM.
     type VMEdge: edge_shape::Edge;
+    type VMMemorySlice: edge_shape::MemorySlice<Edge = Self::VMEdge>;
```

Comment:
This may not be ideal if the VM has the different layout for their array (memory slice). If a VM has struct arrays, the array layout would vary based on the struct layout, such as `[{object ref, usize}]` vs `[{object ref, usize, usize}]`. But maybe they can deal with this in their implementation.

I think we can add a empty implementation of `MemorySlice` for VMs that do not use the memory copy API, so they could just use the empty impl. If they use the empty impl, and ever call the memory copy API, the impl will panic.

Reply:
> But maybe they can deal with this in their implementation.

Yes. For this case the VM is required to filter out invalid pointers on memory iteration. I've put this in the comments.

> I think we can add a empty implementation of `MemorySlice`

Done